### PR TITLE
diff and bookmark enhancements

### PR DESCRIPTION
**Pull request by ahrens**
### Description

This PR includes 3 related features that work well together:

`zfs diff -a` shows which specific blocks were modified

`zfs diff` from a bookmark (but it can't show renamed files)

`zfs bookmark` from a filesystem, creating a bookmark which represents current point in time.  Not useful for `zfs send`, but can be used with `zfs diff`.

### Motivation and Context

This makes `zfs diff` useful in more situations.  For example, to find which blocks in a database or VDI file were changed.

### How Has This Been Tested?

Manual testing only at this point.  I'd like to add test cases to the test suite, but there aren't any tests for "zfs diff" at all, so it seems strange to add tests for just the new functionality I'm adding.  I'm open to input on what should be required for this PR.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by loli10K** [comment]
May be a suitable fix for https://github.com/zfsonlinux/zfs/issues/2470

File: man/man8/zfs.8 [hunk 1]
```diff
@@ -3928,27 +3940,45 @@ descendent file systems.
...
-The first column is a character indicating the type of change, the other columns
-indicate pathname, new pathname
+Display the differences in a filesystem between two points in time.
+The earlier point in time is specified as the first argument, which can be a
+snapshot or a bookmark.
+The later point in time is specified by the second argument, which can be a
+snapshot, a filesystem, or absent.
+If the second argument is absent, it is assumed to be the current state of
+the first argument's filesystem.
+.sp
```
Comment (behlendorf) [inline man/man8/zfs.8]: You're going to want to replace `.sp` with `.Pp` for paragraph breaks everywhere.

File: module/zfs/dmu_diff.c [hunk 2]
```diff
@@ -81,29 +83,16 @@ report_free_dnode_range(struct diffarg *da, uint64_t first, uint64_t last)
 	return (0);
 }
 
-static int
-report_dnode(struct diffarg *da, uint64_t object, dnode_phys_t *dnp)
+static inline boolean_t
+overflow_multiply(uint64_t a, uint64_t b, uint64_t *c)
```
Comment (behlendorf) [inline module/zfs/dmu_diff.c]: Both gcc and clang provide a `__builtin_mul_overflow()` to reliably check for overflows.

https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html
https://clang.llvm.org/docs/LanguageExtensions.html#checked-arithmetic-builtins

File: module/zfs/dsl_bookmark.c [hunk 3]
```diff
@@ -120,15 +121,19 @@ dsl_bookmark_create_check_impl(dsl_dataset_t *snapds, const char *bookmark_name,
...
-		return (SET_ERROR(EINVAL));
-
 	error = dsl_bookmark_hold_ds(dp, bookmark_name,
 	    &bmark_fs, FTAG, &shortname);
 	if (error != 0)
 		return (error);
 
-	if (!dsl_dataset_is_before(bmark_fs, snapds, 0)) {
+	if (!dsl_dataset_is_snapshot(snapds) &&
+	    snapds != bmark_fs) {
```
Comment (behlendorf) [inline module/zfs/dsl_bookmark.c]: nit: looks like it fits on one line.

File: module/zfs/zfs_ioctl.c [hunk 4]
```diff
@@ -5355,6 +5355,7 @@ zfs_ioc_tmp_snapshot(zfs_cmd_t *zc)
  * zc_name		name of "to" snapshot
  * zc_value		name of "from" snapshot
  * zc_cookie		file descriptor to write diff data on
+ * zc_flags		set if blockwise diff is requested
```
Comment (behlendorf) [inline module/zfs/zfs_ioctl.c]: Let's not commit all 32-bits to the blockwise diff flag and instead mask 0x1 for this purpose.

**Comment by behlendorf** [comment]
@loli10K was nice enough to write some `zfs diff` test cases for the existing functionality, #6686.  The new functionality introduced in this PR should be able to build on those tests.

**Comment by arshad512** [comment]
Hi @behlendorf, @ahrens - Please let me know if I can pick this ticket.  It looks like good feature to add and nobody seems to be working on it.

**Comment by ahrens** [comment]
@arshad512 I'd be happy for you to pick this up, let me know if I can answer any questions for you.

**Comment by arshad512** [comment]
> @arshad512 I'd be happy for you to pick this up, let me know if I can answer any questions for you.

Thanks @ahrens !  While I am getting your patch to compile on latest master. How you want the the output to end-user be available ?

For Example: While I can get the idea of difference while doing the `zfs list -t all` (how large is the difference)

`# echo "110" > /myfirstpool/mydataset01/TEST110`

```
[root@mrpel7 zfs]# zfs list -t all
NAME                                 USED  AVAIL     REFER  MOUNTPOINT
myfirstpool                         1.41M  62.6M      666K  /myfirstpool
myfirstpool/mydataset01              682K  62.6M      668K  /myfirstpool/mydataset01
myfirstpool/mydataset01@snapshot01    14K      -      666K  -   
```

How do you vision the output of blocks modified be displayed here  ? 

```
[root@mrpel7 zfs]# zfs diff myfirstpool/mydataset01@snapshot01 myfirstpool/mydataset01
+	/myfirstpool/mydataset01/TEST110
M	/myfirstpool/mydataset01/
```

**Comment by ahrens** [comment]
@arshad512 
> How you want the the output to end-user be available ?

Are you suggesting that we should change the way that the PR currently outputs the changed blocks to the end user, using the `W` (write) and `H` (hole) indicators, as documented in the manpage?

> I can get the idea of difference while doing the zfs list -t all (how large is the difference)

The best way to see the magnitude of differences is with the `written` property, specifically `zfs get written@oldsnap pool/fs@newsnap`.

**Comment by arshad512** [comment]
Hi @ahrens 
> Are you suggesting that we should change the way that the PR currently outputs the changed blocks to the end user, using the `W` (write) and `H` (hole) indicators, as documented in the manpage?

Sorry my question did not came out correct. _Initially_ I thought the block difference would be cumulative for all changes. I was clearly incorrect. I get it now. 

> 
> > I can get the idea of difference while doing the zfs list -t all (how large is the difference)
> 
> The best way to see the magnitude of differences is with the `written` property, specifically `zfs get written@oldsnap pool/fs@newsnap`.

Right! - This question was with the same incorrect assumption above.  I get it now.

**Comment by ahrens** [comment]
I noticed that this PR hasn’t been updated in some time.  We would like to see this feature added, but it seems that it isn’t quite complete, so we’re going to close it for now.

@arshad512, if you'd like to pick this up, please do so and open a new PR.

**Comment by arshad512** [comment]
> I noticed that this PR hasn’t been updated in some time. We would like to see this feature added, but it seems that it isn’t quite complete, so we’re going to close it for now.
> 
> @arshad512, if you'd like to pick this up, please do so and open a new PR.

Hi @ahrens, I intended to work on this, however got little busy. It is in my plan to pick this and complete. I get to this soon. Thanks!

**Comment by ahrens** [comment]
@arshad512 Great, look forward to reviewing it!

**Pull request closed without merging**